### PR TITLE
Switch to prometheus

### PR DESCRIPTION
**Pull request by overvenus**
This is the beginning of switching to prometheus.

TODO:
 - [x] `metric_inc!()`
 - [x] `metric_count!()`
 - [x] `metric_decr!()`
 - [x] `metric_time!()`
 - [x] `metric_gauge!()`
 - [x] `metric_mark!()`
 - [x] `metric_meter!()`.

Please leave a comment if you find something wrong.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -49,6 +49,9 @@ bytes  = "0.3.0"
 nix = "0.6.0"
 utime = "0.1"
 chrono = "0.2"
+prometheus = "0.2.2"
```
Comment (siddontang) [inline Cargo.toml]: use 0.2.3

File: Cargo.toml [hunk 2]
```diff
@@ -49,6 +49,9 @@ bytes  = "0.3.0"
...
+lazy_static = "0.2.1"
+hyper = "0.9"
```
Comment (siddontang) [inline Cargo.toml]: As we discussed before, we don't want to add another port now. so now the best way is to logging metrics periodically, like 5m. 

/cc @ngaut

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -750,6 +751,22 @@ fn main() {
...
     util::print_tikv_info();
 
+    // Run prometheus client.
+    let prometheus_addr = get_string_value("",
+                                           "metric.prometheus",
+                                           &matches,
+                                           &config,
+                                           Some(String::new()),
+                                           |v| v.as_str().map(|s| s.to_owned()));
+    let prometheus_handler = match prometheus_addr.len() {
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: I think we don't need to join this thread, we can let it run forever until program ends.

File: etc/config-template.toml [hunk 4]
```diff
@@ -35,6 +35,7 @@ end-point-concurrency = 8
 addr = ""
 # metric prefix.
 prefix = "tikv"
+prometheus = "127.0.0.1:8989"
```
Comment (siddontang) [inline etc/config-template.toml]: do you want to support push now?

File: etc/config-template.toml [hunk 5]
```diff
@@ -35,6 +35,7 @@ end-point-concurrency = 8
 addr = ""
 # metric prefix.
 prefix = "tikv"
```
Comment (siddontang) [inline etc/config-template.toml]: remove the prefix config

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -330,19 +331,29 @@ impl Peer {
     fn send_ready_metric(&self, ready: &Ready) {
         if !ready.messages.is_empty() {
             metric_count!("raftstore.send_raft_message", ready.messages.len() as i64);
+            PEER_RAFT_READY_COUNTER_VEC.with_label_values(&["send raft message"])
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: seem a little long for the label value
maybe we can use message, committed, append and snapshot directly?

File: src/raftstore/store/worker/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use prometheus::CounterVec;
+
+lazy_static! {
+    pub static ref PD_REQ_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_raftstore_pd_request_sent_total",
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics.rs]: I think using pd for the subsystem is enough.

File: etc/config-template.toml (hunk 5 above)
Comment (overvenus) [inline etc/config-template.toml]: Later, I will remove all statsd things at once.

File: src/raftstore/store/worker/metrics.rs (hunk 7 above)
Comment (overvenus) [inline src/raftstore/store/worker/metrics.rs]: I don't think so, tikv has a mod named `pd`, I think using `pd`, as a subsystem, is better for that mod.

File: src/util/metric/mod.rs [hunk 8]
```diff
@@ -160,3 +162,22 @@ impl MetricSink for BufferedUdpMetricSink {
...
+    loop {
+        let metric_familys = prometheus::gather();
+        encoder.encode(&metric_familys, &mut buffer).unwrap();
+
+        // Output to the standard output.
+        info!("{}", String::from_utf8(buffer.clone()).unwrap());
+
+        buffer.clear();
+        thread::sleep(log_interval);
+        buffer.clear();
```
Comment (siddontang) [inline src/util/metric/mod.rs]: no need to clear again.

File: src/util/metric/mod.rs [hunk 9]
```diff
@@ -160,3 +162,22 @@ impl MetricSink for BufferedUdpMetricSink {
         self.append_to_buffer(metric)
     }
 }
+
+/// `run_prometheus` runs a prometheus client, it blocks current thread.
+pub fn run_prometheus() {
+    let encoder = TextEncoder::new();
+    let mut buffer = Vec::<u8>::new();
+
+    let log_interval = Duration::from_secs(1);
```
Comment (siddontang) [inline src/util/metric/mod.rs]: use a config for this.

**Comment by overvenus** [comment]
PTAL @siddontang @BusyJay @disksing 

Update Cargo.lock too?

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -464,7 +470,7 @@ impl Peer {
         }
 
         debug!("{} propose command with uuid {:?}", self.tag, cmd.uuid);
-        metric_incr!("raftstore.propose");
+        PEER_PROPOSAL_COUNTER_VEC.with_label_values(&["all", "all"]).inc();
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: why two "all"?

Comment (overvenus) [inline src/raftstore/store/peer.rs]: All types and status of proposal.

File: src/raftstore/store/worker/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use prometheus::{CounterVec, HistogramVec, Histogram};
+
+lazy_static! {
+    pub static ref PD_REQ_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_pd_request_sent_total",
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics.rs]: tikv_raftstore_pd

File: src/server/coprocessor/endpoint.rs [hunk 12]
```diff
@@ -195,10 +196,13 @@ impl TiDbEndPoint {
 impl TiDbEndPoint {
     fn handle_requests(&self, reqs: Vec<RequestTask>) {
         for t in reqs {
-            let timer = SlowTimer::new();
-            let tp = t.req.get_tp();
+            let histogram =
+                COPR_REQ_HISTOGRAM_VEC.with_label_values(&[&format!("{}", t.req.get_tp())]);
```
Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: can we use `get_tp` directly?

Comment (overvenus) [inline src/server/coprocessor/endpoint.rs]: But `with_label_values` takes `&[&str]`.

Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: `&[get_tp()]` doesn't work?

Comment (overvenus) [inline src/server/coprocessor/endpoint.rs]: yep

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

File: src/bin/tikv-server.rs [hunk 13]
```diff
@@ -750,6 +711,15 @@ fn main() {
     // Print version information.
     util::print_tikv_info();
 
+    // Run prometheus client.
+    let report_interval = get_integer_value("",
+                                            "metric.report-interval",
+                                            &matches,
+                                            &config,
+                                            Some(1_000),
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: 1s is too short. Use 5m.

**Comment by siddontang** [comment]
PTAL @disksing

File: src/raftstore/store/worker/metrics.rs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+
+use prometheus::{CounterVec, HistogramVec, Histogram};
+
+lazy_static! {
+    pub static ref PD_REQ_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_raftstore_pd_request_sent_total",
+            "Total number of pd client request sent.",
+            &["type", "stauts"]
```
Comment (ngaut) [inline src/raftstore/store/worker/metrics.rs]: s/stauts/status/

**Comment by ngaut** [comment]
LGTM

**Pull request merged**